rust/src/util.rs

```diff
@@ -84,13 +84,67 @@ impl Display for Sizing {
 }
 
 #[derive(Clone, Eq, Debug)]
-#[cfg_attr(feature = "serde", derive(Serialize, Deserialize), serde(crate = "serde_crate"))]
 pub struct Variant {
     pub name: VariantName,
     pub tag: u8,
 }
 impl_strict_struct!(Variant, STRICT_TYPES_LIB; name, tag);
 
+#[cfg(feature = "serde")]
+mod _serde {
+    use std::str::FromStr;
+
+    use serde_crate::ser::SerializeStruct;
+    use serde_crate::{Deserialize, Deserializer, Serialize, Serializer};
+
+    use super::*;
+
+    impl Serialize for Variant {
+        fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+        where S: Serializer {
+            if serializer.is_human_readable() {
+                serializer.serialize_str(&format!("{}:{}", self.name, self.tag))
+            } else {
+                let mut s = serializer.serialize_struct("Variant", 2)?;
+                s.serialize_field("name", &self.name)?;
+                s.serialize_field("tag", &self.tag)?;
+                s.end()
+            }
+        }
+    }
+
+    impl<'de> Deserialize<'de> for Variant {
+        fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+        where D: Deserializer<'de> {
+            if deserializer.is_human_readable() {
+                let s = String::deserialize(deserializer)?;
+                let parts: Vec<&str> = s.split(':').collect();
+                if parts.len() != 2 {
+                    return Err(serde::de::Error::custom(format!(
+                        "Invalid variant format: '{}'. Expected 'name:tag'",
+                        s
+                    )));
+                }
+                let name = VariantName::from_str(parts[0]).map_err(|e| {
+                    serde::de::Error::custom(format!("Invalid variant name: {}", e))
+                })?;
+                let tag = parts[1]
+                    .parse::<u8>()
+                    .map_err(|e| serde::de::Error::custom(format!("Invalid variant tag: {}", e)))?;
+                Ok(Variant { name, tag })
+            } else {
+                #[cfg_attr(feature = "serde", derive(Deserialize), serde(crate = "serde_crate"))]
+                struct VariantFields {
+                    name: VariantName,
+                    tag: u8,
+                }
+                let VariantFields { name, tag } = VariantFields::deserialize(deserializer)?;
```

Comment:
This will probably not work with a tagged YAML, since above in serializer you set explicit name `Variant` for serialization - and here by deriving deserializer it sets the name to `VariantFields`.

Reply:
I think you can add `serde(rename = "Variant")` and this should solve the issue

Reply:
Ok, done